x-pack/plugins/apm/server/lib/services/get_service_transaction_groups_statistics.ts

```diff
@@ -0,0 +1,179 @@
+  serviceName,
+  transactionNames,
+  setup,
+  numBuckets,
+  searchAggregatedTransactions,
+  transactionType,
+  latencyAggregationType,
+}: {
+  serviceName: string;
+  transactionNames: string[];
+  setup: Setup & SetupTimeRange;
+  numBuckets: number;
+  searchAggregatedTransactions: boolean;
+  transactionType: string;
+  latencyAggregationType: LatencyAggregationType;
+}): Promise<
+  Record<
+    string,
+    {
+      latency: Coordinate[];
+      throughput: Coordinate[];
+      errorRate: Coordinate[];
+      impact: number;
+    }
+  >
+> {
+  const { apmEventClient, start, end, esFilter } = setup;
+  const { intervalString } = getBucketSize({ start, end, numBuckets });
+
+  const field = getTransactionDurationFieldForAggregatedTransactions(
+    searchAggregatedTransactions
+  );
+
+  const response = await apmEventClient.search({
+    apm: {
+      events: [
+        getProcessorEventForAggregatedTransactions(
+          searchAggregatedTransactions
+        ),
+      ],
+    },
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            { term: { [SERVICE_NAME]: serviceName } },
+            { term: { [TRANSACTION_TYPE]: transactionType } },
+            { range: rangeFilter(start, end) },
+            ...getDocumentTypeFilterForAggregatedTransactions(
+              searchAggregatedTransactions
+            ),
+            ...esFilter,
+          ],
+        },
+      },
+      aggs: {
+        total_duration: { sum: { field: TRANSACTION_DURATION } },
+        transaction_groups: {
+          terms: {
+            field: TRANSACTION_NAME,
+            include: transactionNames,
+            size: transactionNames.length,
+          },
+          aggs: {
+            transaction_group_total_duration: {
+              sum: { field: TRANSACTION_DURATION },
+            },
+            timeseries: {
+              date_histogram: {
+                field: '@timestamp',
+                fixed_interval: intervalString,
+                min_doc_count: 0,
+                extended_bounds: {
+                  min: start,
+                  max: end,
+                },
+              },
+              aggs: {
+                throughput_rate: {
+                  rate: {
+                    field: TRANSACTION_DURATION,
+                    unit: 'minute',
+                    mode: 'value_count',
+                  },
+                },
+                ...getLatencyAggregation(latencyAggregationType, field),
+                [EVENT_OUTCOME]: {
+                  filter: {
+                    term: { [EVENT_OUTCOME]: EventOutcome.failure },
+                  },
+                },
+              },
+            },
+          },
+        },
+      },
+    },
+  });
+
+  const buckets = response.aggregations?.transaction_groups.buckets ?? [];
+
+  const totalDuration = response.aggregations?.total_duration.value;
+
+  return buckets.reduce((acc, bucket) => {
```

Comment:
nit: We could also map first, and then use `keyBy` after mapping. `.reduce` is always a little harder to follow.

Reply:
Actually, I think it's easier to have this API return an array, and have the client take care of making sure that data is merged properly. I think the tests would look cleaner.

Reply:
Why do you think that? I think it makes sense for this API to return an object with all groupNames it received.

Reply:
If you look at your API test, you're doing a lot of conversions from a map to an array (by using Object.values or Object.keys). That might be a good indicator that the data structure is not optimal.

Reply:
I changed the tests, I think the API should still return an object.